### PR TITLE
introduce utf8 helper

### PR DESCRIPTION
**Pull request by janbar**
Hi, it is a way to resolve issue caused by conversions from utf8 to localized wchar (upper/lower), in particular for location service (search). There is no longer dependency with system locale to handle operations on unicode string, similary using libICU.
The characters map has been generated from UTF8. 8.0, and I include all codepoint related to Latin, European languages + armenian + georgian and some others known to have upper/lower matching. The others codepoint are simply pass throught.
The code is fully static with minimal footprint. For instance it provides 3 functions impacted by the issue ( upper, lower, normForLookup ). Also now the function "normalize" remove duplicate spaces and breakers.
It just for preview, surely need rework.

### CODE REVIEW

File: libosmscout/include/osmscout/util/utf8helper_charmap.h [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+{
+
+constexpr char None         = 0x00;
+constexpr char IsSpace      = 0x01;
+constexpr char IsBreaker    = 0x02;
+constexpr char IsControl    = 0x04;
+
+using byte = unsigned char;
+
+using codepoint = unsigned int;
```
Comment (Framstag) [inline libosmscout/include/osmscout/util/utf8helper_charmap.h]: Shouldn't you not better use something like uint8_t...uint32_t?

File: libosmscout/src/osmscout/util/utf8helper.cpp [hunk 2]
```diff
@@ -0,0 +1,276 @@
+#include <osmscout/util/utf8helper.h>
+#include <osmscout/util/utf8helper_charmap.h>
+
+#include <string>
+#include <cstring>
+
+namespace utf8helper
+{
+
+static std::string parse(const std::string& text, codepoint (*func)(const character*, char context)){
```
Comment (Framstag) [inline libosmscout/src/osmscout/util/utf8helper.cpp:10]: It would be nice to have some description for what the code is doing. Also some external link to a more in deep documentation would be helpful.

What about some tests, especially for invalid input to make sure that corner cases are handled correctly?

**Comment by Framstag** [comment]
Reviewed, please give me a hint, if I should merge.

**Comment by Framstag** [comment]
Also, Visual Studio breaks on your code.

File: libosmscout/include/osmscout/util/utf8helper_charmap.h (hunk 1 above)
Comment (janbar) [reply libosmscout/include/osmscout/util/utf8helper_charmap.h]: Yes It would be better. It is an old habit to avoid compiler failure with old platforms. To be fixed including stdint.

**Comment by janbar** [comment]
@framstag, I have to fix the issue you had with Visual Studio, and add some code for test. I will let you know when it will ready.

**Comment by Framstag** [comment]
@janbar Take your time. I should stress also, that this is a great new feature! Thanks for implementing it.

**Comment by janbar** [comment]
@Framstag , I fixed the issue with the microsoft compiler. Seems `char[]` is not char* on this platform. Anyway I made it better by testing with windows. Still to make some primitives to test/stress it. As the utf8 standard is backward compatible with ascii/iso8, you can make it eat a binary file without get an issue. The best effort is to maintain the charmaps to do what we need, i.e generate efficient tokens with transliterate. Probably removing punctuation etc.

**Comment by janbar** [comment]
@Framstag , I have no more change to add. I tested this successfully with android app and now searching works as expected, and it should work anywhere for any body (locale).
About performance it processes more than 25GB per second per core i7, so much more than iconv. Also I tried to design it as cpp class, but the overhead is 25%. IMO no need for overkill engineering, it is OK as is.
